File: rust/examples/linux/other/mpr121/minimal/src/main.rs
use linux_embedded_hal::{Delay, I2cdev};
use periph::chips::other::Mpr121Minimal;
use std::thread::sleep;
use std::time::Duration;

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x5A);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus"); // Open I²C bus, (/dev/i2c-N) → I2cdev
    let mut delay = Delay;
    let mut chip = Mpr121Minimal::new(dev, addr, &mut delay).expect("init");       // Construct MPR121 Minimal, (i2c, addr=0x5A, delay) → Result<Mpr121Minimal, _>
                                                                                   // resets, applies default thresholds (T=12, R=6), enters Run Mode on all 12 electrodes

    loop {
        let t = chip.touched().expect("touched");                                    // Read 12-bit touch bitmask, () → Result<u16, _> bitmask
                                                                                   // bit n=1 means ELEn is currently touched
        println!("touched=0x{:03X}", t);
        sleep(Duration::from_secs(1));
    }
}

File: cpp/examples/linux/other/Mpr121/minimal/main.cpp
#include <cstdio>
#include <cstdlib>
#include <unistd.h>
#include "I2CConnectionLinux.h"
#include "Mpr121.h"

int main() {
    const char* bus_env  = getenv("I2C_BUS");
    const char* addr_env = getenv("I2C_ADDR");
    int     bus  = bus_env  ? atoi(bus_env)       : 1;
    uint8_t addr = addr_env ? (uint8_t)strtol(addr_env, nullptr, 0) : 0x5A;
    I2CConnectionLinux connection(bus, addr);                                 // Create I2C connection, (bus, addr) → I2CConnectionLinux

    MPR121Minimal mpr(connection);                                             // Create MPR121 Minimal, (connection) → MPR121Minimal

    while (true) {
        uint16_t t = mpr.touched();                                            // Read 12-bit touch bitmask, () → uint16_t bitmask
        printf("touched=0x%03X\n", t);
        usleep(1000 * 1000);
    }
    return 0;
}
